### ISSUE TITLE
Group-2:  Implement K-fold Cross-Validation Harness And Appropriate Score Evaluation Metric

### ISSUE THREAD

**Issue by project-delphi** [ccm-bioinfo/Camda24_resistance#10]
**Objective**:
Build an ensembling method using stacking for multivariate output, selecting an appropriate scoring function for cross-validation to evaluate model performance.

#### Tasks to Complete

1. **Define Base Models for Stacking**
   - [ ] Select and define multiple base models (e.g., `RandomForest`, `GradientBoosting`, `KNN`) to be used in the stacking ensemble.
   - [ ] Ensure that the base models are suited for handling multivariate outputs.
   - [ ] Implement the base models in code, and ensure they are ready for integration into the stacking ensemble.

2. **Implement Stacking Ensemble**
   - [ ] Use `sklearn`'s `StackingRegressor` or `StackingClassifier` (depending on the task) to implement stacking.
   - [ ] Define a meta-model (e.g., `LinearRegression`, `XGBoost`) that will take the predictions of base models as input and predict the final output.
   - [ ] Ensure that the meta-model can handle multivariate output.
   - [ ] Test the stacking ensemble to ensure that it functions correctly and outputs multivariate predictions.

3. **Choose a Score Function for Cross-Validation**
   - [ ] Identify an appropriate scoring function for evaluating the performance of the ensemble on multivariate output (e.g., `mean_squared_error` for regression, `accuracy` for classification).
   - [ ] If necessary, implement a custom scoring function that accounts for multivariate outputs.
   - [ ] Ensure that the score function can be used with `cross_val_score` or `cross_validate` from `sklearn`.

4. **Cross-Validation with Stacking Ensemble**
   - [ ] Implement cross-validation using `cross_val_score` or `cross_validate` from `sklearn` to evaluate the performance of the stacking ensemble.
   - [ ] Ensure that the selected score function is applied during cross-validation.
   - [ ] Report the performance metrics (e.g., mean score, standard deviation) of the ensemble across the cross-validation folds.

5. **Testing and Validation**
   - [ ] Test the ensemble method on a sample multivariate dataset to verify the implementation.
   - [ ] Ensure the cross-validation results are consistent and reliable across different folds.
   - [ ] Optimize the stacking method by tuning hyperparameters of base models and the meta-model.

#### Example Code Structure

```python
from sklearn.ensemble import StackingRegressor
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import make_scorer, mean_squared_error

# Define base models
base_models = [
    ('rf', RandomForestRegressor(n_estimators=100)),
    ('gb', GradientBoostingRegressor(n_estimators=100))
]

# Define meta-model
meta_model = LinearRegression()

# Create stacking ensemble
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model)

# Define scoring function (mean squared error for regression)
scoring = make_scorer(mean_squared_error, multioutput='uniform_average')

# Perform cross-validation
scores = cross_val_score(stacking_model, X, y, scoring=scoring, cv=5)
print('Cross-validation scores:', scores)
print('Mean score:', scores.mean())

**Comment by project-delphi** [comment]
reset to clean modules with tested code.
